crates/rattler_repodata_gateway/src/fetch/jlap/mod.rs

```diff
@@ -0,0 +1,711 @@
+                if !patches.is_empty() {
+                    Ok((metadata, patches))
+                } else {
+                    Err(JLAPError::NoPatchesFoundError)
+                }
+            }
+            Err(error) => Err(error),
+        }
+    } else {
+        Err(JLAPError::NoPatchesFoundError)
+    }
+}
+
+/// Applies JLAP patches to a `repodata.json` file
+///
+/// This is a multi-step process that involves:
+///
+/// 1. Opening and parsing the current repodata file
+/// 2. Applying patches to this repodata file
+/// 3. Saving this repodata file to disk
+/// 4. Generating a new `blake2b` hash
+///
+/// The return value is the `blake2b` hash we used to verify the updated file's contents.
+pub async fn apply_jlap_patches(
+    patches: &Vec<&Patch>,
+    repo_data_path: &Path,
+) -> Result<Output<Blake2b256>, JLAPError> {
+    // Open and read the current repodata into a JSON doc
+    let repo_data_contents = match tokio::fs::read_to_string(repo_data_path).await {
+        Ok(contents) => contents,
+        Err(error) => return Err(JLAPError::FileSystemError(error)),
+    };
+
+    let mut doc = match serde_json::from_str(&repo_data_contents) {
+        Ok(doc) => doc,
+        Err(error) => return Err(JLAPError::JSONParseError(error)),
+    };
+
+    // Apply the patches we current have to it
+    for patch in patches {
+        match json_patch::patch(&mut doc, &patch.patch) {
+            Ok(_) => (),
+            Err(error) => return Err(JLAPError::JSONPatchError(error)),
+        }
+    }
+
+    // Save the updated repodata JSON doc
+    let mut updated_file = match tokio::fs::File::create(repo_data_path).await {
+        Ok(file) => file,
+        Err(error) => return Err(JLAPError::FileSystemError(error)),
+    };
+
+    let mut updated_json = match serde_json::to_string_pretty(&doc) {
+        Ok(value) => value,
+        Err(error) => return Err(JLAPError::JSONParseError(error)),
+    };
+
+    // We need to add an extra newline character to the end of our string so the hashes match 🤷‍
+    updated_json.insert(updated_json.len(), '\n');
+    let content = updated_json.into_bytes();
+
+    match updated_file.write_all(&content).await {
+        Ok(_) => Ok(compute_bytes_digest::<Blake2b256>(content)),
+        Err(error) => Err(JLAPError::FileSystemError(error)),
+    }
+}
+
+/// Finds the index of the of the most applicable patch to use
+fn find_current_patch_index(patches: &[Patch], hash: Output<Blake2b256>) -> Option<usize> {
+    for (idx, patch) in patches.iter().enumerate() {
+        if hash == patch.from.unwrap_or_default() {
+            return Some(idx);
+        }
+    }
+
+    None
+}
+
+/// Finds the initial vector in the JLAP response; this is the first line of a full request
+fn get_initial_vector(jlap_response: &str) -> Option<String> {
+    let parts: Vec<&str> = jlap_response.split('\n').collect();
+
+    if !parts.is_empty() {
+        Some(parts[0].to_string())
+    } else {
+        None
+    }
```

Comment:
This avoids an extra allocation:

```suggestion
    jlap_response.split('\n').next().map(Into::into)   
```